plugins/llvm/src/main/java/org/qbicc/plugin/llvm/LLVMCallSiteInfo.java

```diff
@@ -0,0 +1,36 @@
+package org.qbicc.plugin.llvm;
+
+import org.qbicc.context.AttachmentKey;
+import org.qbicc.context.CompilationContext;
+import org.qbicc.graph.Node;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+public class LLVMCallSiteInfo {
+    // This is a rough estimate based on number of callsites for an app with empty main().
+    // A better way would be to collect this stat during compilation and use it to initialize the list capacity.
+    private static final int INITIAL_LIST_SIZE = 100000;
+    private final ArrayList<Node> nodeList = new ArrayList<>(INITIAL_LIST_SIZE);
+    public static final AttachmentKey<LLVMCallSiteInfo> KEY = new AttachmentKey<>();
+
+    private LLVMCallSiteInfo() {}
+
+    public static LLVMCallSiteInfo get(CompilationContext ctxt) {
+        return ctxt.computeAttachmentIfAbsent(KEY, LLVMCallSiteInfo::new);
+    }
+
+    public void mapSpToNode(int spId, Node node) {
+        synchronized (nodeList) {
+            if (spId >= nodeList.size()) {
+                for (int i = nodeList.size(); i <= spId; i++) {
+                    nodeList.add(i, null);
+                }
+            }
```

Comment:
`ensureCapacity` won't work because the elements in the list may be added in out-of-order manner. This makes it necessary to use `ArrayList#set` method which will throw `IndexOutOfBounds` if the size of the list is less than the index being set. Note that `ensureCapacity` does not alter the size of the list, only increases the capacity.